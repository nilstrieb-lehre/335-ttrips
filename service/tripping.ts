// Helpers about computing properties about the current trip.

import { Connection, Stop } from "./transport";

export function nextStop(connection: Connection): Stop | undefined {
  const currentTime = new Date();

  if (
    connection.from.departure &&
    new Date(connection.from.departure) > currentTime
  ) {
    return connection.from;
  }

  return connection.sections.find((section) => {
    const expectedArrival = section.arrival.arrival;
    return expectedArrival && new Date(expectedArrival) > currentTime;
  })?.arrival;
}

export type LatLong = { latitude: number; longitude: number };

// Calculates the angle the user has to walk to from their position (S) to the target (T).
export function targetAngle(self: LatLong, target: LatLong): number {
  const relativeTargetX = target.longitude - self.longitude;
  const relativeTargetY = target.latitude - self.latitude;
  return Math.atan2(relativeTargetX, relativeTargetY) * (180 / Math.PI);
}

const winterthur = { latitude: 47.4970697, longitude: 8.718835 };
const richterswil = { latitude: 47.208373, longitude: 8.707486 };
