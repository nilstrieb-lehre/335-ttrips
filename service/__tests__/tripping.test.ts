import davos_frauenkirch_geneve_rive from "./davos_frauenkirch-genÃ¨ve_rive.json";
import { tripToStopStations } from "../tripping";

it("should convert the connection to a station list", () => {
  const [stations] = tripToStopStations(davos_frauenkirch_geneve_rive);

  expect(stations).toHaveLength(9);
  expect(stations).toMatchSnapshot();
});
