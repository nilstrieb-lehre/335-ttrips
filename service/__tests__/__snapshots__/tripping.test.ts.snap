// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should convert the connection to a station list 1`] = `
[
  {
    "arrival": null,
    "arrivalPlatform": null,
    "departure": 2023-07-14T12:36:00.000Z,
    "departurePlatform": null,
    "station": {
      "coordinate": {
        "type": "WGS84",
        "x": 46.766596,
        "y": 9.799247,
      },
      "distance": null,
      "id": "8509075",
      "name": "Davos Frauenkirch",
      "score": null,
    },
  },
  {
    "arrival": 2023-07-14T12:38:00.000Z,
    "arrivalPlatform": null,
    "departure": 2023-07-14T12:38:00.000Z,
    "departurePlatform": null,
    "station": {
      "coordinate": {
        "type": "WGS84",
        "x": 46.76702,
        "y": 9.79916,
      },
      "distance": null,
      "id": "8581178",
      "name": "Davos Frauenkirch, Landhaus",
      "score": null,
    },
  },
  {
    "arrival": 2023-07-14T12:56:00.000Z,
    "arrivalPlatform": null,
    "departure": 2023-07-14T12:56:00.000Z,
    "departurePlatform": null,
    "station": {
      "coordinate": {
        "type": "WGS84",
        "x": 46.809753,
        "y": 9.84176,
      },
      "distance": null,
      "id": "8574848",
      "name": "Davos Dorf, Bahnhof",
      "score": null,
    },
  },
  {
    "arrival": 2023-07-14T12:58:00.000Z,
    "arrivalPlatform": null,
    "departure": 2023-07-14T13:06:00.000Z,
    "departurePlatform": null,
    "station": {
      "coordinate": {
        "type": "WGS84",
        "x": 46.809718,
        "y": 9.842113,
      },
      "distance": null,
      "id": "8509072",
      "name": "Davos Dorf",
      "score": null,
    },
  },
  {
    "arrival": 2023-07-14T14:13:00.000Z,
    "arrivalPlatform": "5AB",
    "departure": 2023-07-14T14:18:00.000Z,
    "departurePlatform": "3",
    "station": {
      "coordinate": {
        "type": "WGS84",
        "x": 46.967437,
        "y": 9.554043,
      },
      "distance": null,
      "id": "8509002",
      "name": "Landquart",
      "score": null,
    },
  },
  {
    "arrival": 2023-07-14T15:22:00.000Z,
    "arrivalPlatform": "13",
    "departure": 2023-07-14T15:32:00.000Z,
    "departurePlatform": "32",
    "station": {
      "coordinate": {
        "type": "WGS84",
        "x": 47.377847,
        "y": 8.540502,
      },
      "distance": null,
      "id": "8503000",
      "name": "Zürich HB",
      "score": null,
    },
  },
  {
    "arrival": 2023-07-14T18:18:00.000Z,
    "arrivalPlatform": "3",
    "departure": 2023-07-14T18:18:00.000Z,
    "departurePlatform": null,
    "station": {
      "coordinate": {
        "type": "WGS84",
        "x": 46.210228,
        "y": 6.142435,
      },
      "distance": null,
      "id": "8501008",
      "name": "Genève",
      "score": null,
    },
  },
  {
    "arrival": 2023-07-14T18:23:00.000Z,
    "arrivalPlatform": null,
    "departure": 2023-07-14T18:25:00.000Z,
    "departurePlatform": null,
    "station": {
      "coordinate": {
        "type": "WGS84",
        "x": 46.209751,
        "y": 6.14242,
      },
      "distance": null,
      "id": "8587057",
      "name": "Genève, gare Cornavin",
      "score": null,
    },
  },
  {
    "arrival": 2023-07-14T18:33:00.000Z,
    "arrivalPlatform": null,
    "departure": null,
    "departurePlatform": null,
    "station": {
      "coordinate": {
        "type": "WGS84",
        "x": 46.201881,
        "y": 6.152792,
      },
      "distance": null,
      "id": "8587061",
      "name": "Genève, Rive",
      "score": null,
    },
  },
]
`;
